### PR TITLE
Msc proj: Added new conv_v2 class in layer_v2.py in joey, tested convolution with batch, channels

### PR DESCRIPTION
**Pull request by utkarsh8028**

### CODE REVIEW

**Comment by navjotk** [comment]
Did you mean to commit the `.idea` directory? That's probably your IDE keeping UI state. 
Ditto for the original MNIST data - I don't think we want that to live in our repository. Especially not in two places.

File: Reading_Dataset/read_dataSet.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import numpy as np # linear algebra
```
Comment (navjotk) [inline Reading_Dataset/read_dataSet.py:1]: Move this file to examples/Brats/brats_util.py

Make this a function that accepts dataset path as a parameter.

File: backward_2022.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
+#!/usr/bin/env python
```
Comment (navjotk) [inline backward_2022.py:1]: Move this file to examples/training_lenet.py

Also remove all the excessive comments here.

File: joey/layers_v2.py [hunk 3]
```diff
@@ -0,0 +1,734 @@
+from abc import abstractmethod
```
Comment (navjotk) [inline joey/layers_v2.py:1]: Have you rewritten all layers? Including FullyConnected etc?

File: joey/net.py [hunk 4]
```diff
@@ -68,6 +68,9 @@ def _gen_eqs(self):
 
         input_function = None
 
+        for layer in self._layers:
+            eqs.append(Eq(layer.result, 0))
```
Comment (navjotk) [inline joey/net.py:72]: What is this about?

File: tests/2D_conv_tests/2d_compare_test.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        input.append(temp)
+
+ #to simulate kernel, image with channel 1
+    kernel = [[kernel]]
+    input =[input]
+
+    result_devito = conv_devito_2D(input, kernel, padding,1, print_code = print_results)
+    result_torch = pyTorch_conv(input, kernel, padding,1)
+
+    if print_results:
```
Comment (navjotk) [inline tests/2D_conv_tests/2d_compare_test.py:28]: Might have been useful to write/debug but unnecessary in the long run.

File: tests/2D_conv_tests/2d_compare_test.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+@pytest.mark.parametrize("input_size, kernel, padding", [(5, custom_kernel, 2),(50, custom_kernel, 6) ])
+def test_Devito_pyTorch_padding(input_size, kernel, padding, print_results = False ):
+    c=0;
+    input =[]
+    for i in range(0,input_size):
+        temp =[]
+        for j in range(0,input_size):
+            temp.append(c)
+            c = c+1;
```
Comment (navjotk) [inline tests/2D_conv_tests/2d_compare_test.py:18]: What is going on here?

File: tests/2D_conv_tests/2d_compare_test.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        print ("devito",result_devito)
+
+
+    assert(np.allclose(result_devito, result_torch))
+
+
+
+
+@pytest.mark.parametrize("input_size, kernel, stride", [(5, custom_kernel, 2),(50, custom_kernel, 5) ])
+def test_Devito_pyTorch_stride(input_size, kernel, stride, print_results = False ):
```
Comment (navjotk) [inline tests/2D_conv_tests/2d_compare_test.py:40]: Can we combine this test with the previous one?

File: tests/2D_conv_tests/2d_compare_test.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        print ("torch",result_torch[0])
+
+        print ("devito",result_devito)
+
+    assert(np.allclose(result_devito, result_torch))
+
+
+
+
+custom_kernel1 = [[0.5,0.5,0.5,0.5,0.5] ,[1,0,-1,0.5,0.5],[0.5,0.5,0.5,0.5,0.5],[1,0,-1,0.5,0.5],[0.5,0.5,0.5,0.5,0.5]]
```
Comment (navjotk) [inline tests/2D_conv_tests/2d_compare_test.py:68]: This can become a pytest fixture.

File: tests/2D_conv_tests/2d_compare_test.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+@pytest.mark.parametrize("input_size, kernel, channels, no_kernels, padding, stride", [(5, custom_kernel, 3, 3, 1, 1),(50, custom_kernel1, 2,2, 6, 4) ])
+def test_Devito_pyTorch(input_size,kernel,channels, no_kernels, padding, stride, print_results = False):
```
Comment (navjotk) [inline tests/2D_conv_tests/2d_compare_test.py:71]: Can we combine this test?

File: tests/2D_conv_tests/2d_compare_test.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+
+    print("Do they match", np.allclose(result_devito, result_torch))
+
+    #assert(np.allclose(result_devito, result_torch))
+
+
+@pytest.mark.parametrize("input_size, kernel, channels, no_kernels, padding, stride", 
+[((3,5), custom_kernel, 3, 3, 1, 1),((5,50), custom_kernel1, 2,2, 6, 3), ((5,35), custom_kernel1,5,3, 10, 2) ])
+def test_Joey_pyTorch_Conv(input_size,kernel,channels, no_kernels, padding, stride, print_results = False):
```
Comment (navjotk) [inline tests/2D_conv_tests/2d_compare_test.py:112]: Can we combine?

File: tests/2D_conv_tests/Devito_Conv_2d.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+
+'''
+    Input should be a 3D matrix like 3x512x512 image where 3 is the channels and 512x512 is the size of image.
+    Similary kernel should be be 3D matrix where 3x5x5, 3 is the channel which 
+    should always match the channels in input image, 5x5 is the kernel filter.
+    Assumptioms - kernel would be of square size like 5x5, 3x3, same goes for image
+'''
+
+def conv_devito_2D(input,kernel, padding,stride, print_code = True):
```
Comment (navjotk) [inline tests/2D_conv_tests/Devito_Conv_2d.py:19]: This doesn't seem to be using the implementation from joey.layers_v2. Why is this here?

File: tests/2D_conv_tests/conv.c [hunk 12]
```diff
@@ -0,0 +1,81 @@
+#include<stdio.h>
+
+
+void convolute(double output [3][3], double input[5][5], double kernel[3][3] )
```
Comment (navjotk) [inline tests/2D_conv_tests/conv.c:4]: Why is this here?

File: tests/2D_conv_tests/devito_conv_c_1d.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
+import devito
```
Comment (navjotk) [inline tests/2D_conv_tests/devito_conv_c_1d.py:1]: Why is this here?

File: tests/2D_conv_tests/devito_sample.py [hunk 14]
```diff
@@ -0,0 +1,10 @@
+from devito import Grid, Function, Eq,Operator
```
Comment (navjotk) [inline tests/2D_conv_tests/devito_sample.py:1]: Why is this here?

File: tests/TransConv_test.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
+"""
```
Comment (navjotk) [inline tests/TransConv_test.py:1]: What is going on this file?

File: tests/conv_pad_test.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
+"""
```
Comment (navjotk) [inline tests/conv_pad_test.py:1]: Are you using this?

File: tests/performance_evaluation.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
+"""
```
Comment (navjotk) [inline tests/performance_evaluation.py:1]: Make sure you have a good reason to add any file you do. Why are we adding this here?

File: tests/performance_evaluation_PyTorch.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
+from pickle import NONE
```
Comment (navjotk) [inline tests/performance_evaluation_PyTorch.py:1]: Why?

File: tests/performance_evaluation_joey.py [hunk 19]
```diff
@@ -0,0 +1,121 @@
+import joey
```
Comment (navjotk) [inline tests/performance_evaluation_joey.py:1]: Again, why?

File: unet/README.md [hunk 20]
```diff
@@ -0,0 +1,6 @@
+Independent Research Project - Code
```
Comment (navjotk) [inline unet/README.md:1]: Why?

File: unet/TransConv_test.py [hunk 21]
```diff
@@ -0,0 +1,48 @@
+"""
```
Comment (navjotk) [inline unet/TransConv_test.py:1]: Adding the file twice?

File: unet/additional_layers.py [hunk 22]
```diff
@@ -0,0 +1,243 @@
+"""
```
Comment (navjotk) [inline unet/additional_layers.py:1]: 1. Why are you adding this?
2. Why here?

File: unet/additions_base.py [hunk 23]
```diff
@@ -0,0 +1,51 @@
+"""
```
Comment (navjotk) [inline unet/additions_base.py:1]: Why are you adding this?

File: unet/conv_pad_test.py [hunk 24]
```diff
@@ -0,0 +1,88 @@
+"""
```
Comment (navjotk) [inline unet/conv_pad_test.py:1]: Why are you adding this?

File: unet/performance_evaluation.py [hunk 25]
```diff
@@ -0,0 +1,236 @@
+"""
```
Comment (navjotk) [inline unet/performance_evaluation.py:1]: Why are you adding this?

**Review by navjotk — CHANGES_REQUESTED** [review]
You are adding too many unnecessary things. Please keep changes in a single PR to a minimum. 37 changed files is the largest PR I have seen in many years.

File: joey/layers_v2.py (hunk 3 above)
Comment (utkarsh8028) [reply joey/layers_v2.py:1]: I was planning to after confirming that conv class is okay.. I can circle back here later once I am done with 3D I guess?

File: unet/performance_evaluation.py (hunk 25 above)
Comment (utkarsh8028) [reply unet/performance_evaluation.py:1]: I guess I should'nt have integrated the unet folder within my branch, I thought I can have it here for reference and clean up later, some of the test files you have flagged are from the unet tests.. I will remove this from the branch

File: tests/performance_evaluation.py (hunk 17 above)
Comment (utkarsh8028) [reply tests/performance_evaluation.py:1]: performance_evaluation_joey, performance_evaluation_pyTorch and the driver class performance_evaluation.py is the modified and refactor lenet evaluation test that I performed with the correct loss function to understand the  reason for 'Nan' in the forward pass of the lenet

File: tests/2D_conv_tests/Devito_Conv_2d.py (hunk 11 above)
Comment (utkarsh8028) [reply tests/2D_conv_tests/Devito_Conv_2d.py:19]: I am using the classes in 2D_conv_test to develop and test... These can go away once I am done with 3D conv...

File: joey/net.py (hunk 4 above)
Comment (utkarsh8028) [reply joey/net.py:72]: I think I only made a spacing change in net.py  at line 182, rest of the changes were done by Maks, squashing commits tagged me along side Maks I guess
